### DISCUSSION TITLE
How to render an array of actors selected using `useSelector`?

### DISCUSSION THREAD

**Discussion by Girish21** [statelyai/xstate#3379] [Q&A]
I have an array of actors, and I use `useInterpret`  to create the instance of the machine. And in the component where I want to render the list of actors, I use `useSelector` to select the list of actor references from the context of the machine. I then map them, render a list, pass the actor ref to the component, and use `useActor` to tap into the state of the actor reference, sample sandbox [actor-array](https://codesandbox.io/s/xstate-actor-array-te2fq3). Everything is working as expected, except I get a warning in the console that `the result of getSnapshot should be cached to avoid an infinite loop`. Cool, I can use the third argument of `useSelector` to compare the previous and the next state to check if a rerender is required. Things get weird here when I add the compare function, and the actors are not getting rerendered when the internal states of the actor get changed [actor-array-compare](https://codesandbox.io/s/xstate-actor-array-compare-4qoe2c?file=/src/App.js).

I'm missing something here, and I'm not sure if I'm following the best practices. If anyone can take a look at the example sandboxes and let me know how this should be approached, that would be really great! Thanks!

**Comment by Andarist** [comment]
It's not really your fault - this should just work. However, because of the old architectural choices in XState we have some problems with reusing computed initial states of machines and we just recompute them each time. So in this demo, the problem is in the internal `getSnapshot` used within `useSelector`. We are working around this problem (somewhat) within `useMachine` so perhaps we could do a similar thing here... 🤔 I'll experiment with this after the weekend.

**Reply by jzecca** [reply]
I guess I have the same problem and I'm really interested in any fix or workaround.

I don't think I'm doing anything fancy here, but the first use of the `useChildMachine` hook triggers the warning:

```js
import { createContext, useContext } from 'react';
import { useInterpret, useSelector } from '@xstate/react';
import main from '../machines/main';

const Context = createContext({});

export function MachineProvider({ children }) {
    const service = useInterpret(main);

    return (
        <Context.Provider value={service}>
            {children}
        </Context.Provider>
    );
}

export function useMachine() {
    return useContext(Context);
}

export function useChildMachine(id) {
    const machine = useContext(Context);

    // "Warning: The result of getSnapshot should be cached to avoid an infinite loop"
    return useSelector(machine, state => state.children[id]);
}
```

**Reply by jzecca** [reply]
**_Really ugly_** workaround, temporarily overriding `console.error`:

```js
export function useChildMachine(id) {
    const machine = useContext(Context);

    // Override console.error to suppress the message
    const _error = console.error;
    console.error = (...args) => args[0].includes('getSnapshot') ? undefined : _error(...args);

    const child = useSelector(machine, state => state.children[id]);

    console.error = _error; // restore console.error

    return child;
}
```

**Reply by Andarist** [reply]
Could you tell me how your `main` machine looks like? It would be best if you could always share a runnable repro case for any given reported issue - it allows both sides (you, the reporter, and us, the maintainer) to ensure that we discuss/look at the same thing.

**Reply by jzecca** [reply]
Sorry for the delay, here is a reproducer:
https://github.com/jzecca/xstate-getsnapshot-reproducer

Do not hesitate to tell me if you want me to move this to an issue.

**Reply by Andarist** [reply]
Ok, I think that I see the problem - could you report this as an issue so I don't lose track of it? I'm gonna look into fixing this tomorrow.

**Reply by jzecca** [reply]
Sure, here it is: https://github.com/statelyai/xstate/issues/3497
